README.md
# Aura.Accept

Provides content-negotiation tools using `Accept*` headers.

## Foreword

### Installation

This library requires PHP 5.3 or later; we recommend using the latest available version of PHP as a matter of principle. It has no userland dependencies.

It is installable and autoloadable via Composer as [aura/accept](https://packagist.org/packages/aura/accept).

Alternatively, [download a release](https://github.com/auraphp/Aura.Accept/releases) or clone this repository, then require or include its _autoload.php_ file.

### Quality

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/auraphp/Aura.Accept/badges/quality-score.png?b=develop-2)](https://scrutinizer-ci.com/g/auraphp/Aura.Accept/)
[![Code Coverage](https://scrutinizer-ci.com/g/auraphp/Aura.Accept/badges/coverage.png?b=develop-2)](https://scrutinizer-ci.com/g/auraphp/Aura.Accept/)
[![Build Status](https://travis-ci.org/auraphp/Aura.Accept.png?branch=develop-2)](https://travis-ci.org/auraphp/Aura.Accept)

To run the [PHPUnit](http://phpunit.de/manual/) unit tests at the command line, issue `composer install` and then `vendor/bin/phpunit` at the package root. This requires [Composer](http://getcomposer.org/) to be available as `composer`.

This library attempts to comply with [PSR-1][], [PSR-2][], and [PSR-4][]. If
you notice compliance oversights, please send a patch via pull request.

[PSR-1]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md
[PSR-2]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md
[PSR-4]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader.md

### Community

To ask questions, provide feedback, or otherwise communicate with the Aura community, please join our [Google Group](http://groups.google.com/group/auraphp), follow [@auraphp on Twitter](http://twitter.com/auraphp), or chat with us on #auraphp on Freenode.


## Getting Started

### Instantiation

First, instantiate a _AcceptFactory_ object, then use it to create an _Accept_
object.

```php
<?php
use Aura\Accept\AcceptFactory;

$accept_factory = new AcceptFactory($_SERVER);
$accept = $accept_factory->newInstance();
?>
```

The _Accept_ object provides convenienence methods to negotiate between
acceptable (to the client) and available (from the application) charset,
encoding, language, and media-type values. The methods are:

- `negotiateCharset()`
- `negotiateEncoding()`
- `negotiateLanguage()`
- `negotiateMedia()`

Your application code, knowing what values it has available, should pass an
array of the available values to the `negotiate*()` method. (The values that are
acceptable to the client are already indicated by `$_SERVER`).

The result returned from the method will be a `*Value` object describing the
highest-quality match that was negotiated between the available and acceptable
values. If there is no negotiable match, the result will be `false`.

> N.b. Accept headers can be kind of complicated. See the
> [HTTP Header Field Definitions](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html)
> for more detailed information regarding quality factors, matching rules,
> and parameters extensions.


### Negotiating Media Types

To negotiate a media type (aka a content type), call the `negotiateMedia()`
method with a list of available media types. The available types should be in
the order you prefer for delivery, from "most preferred" to "least preferred".

```php
<?php
// assume the request indicates these Accept values (XML is best, then CSV,
// then anything else)
$_SERVER['HTTP_ACCEPT'] = 'application/xml;q=1.0,text/csv;q=0.5,*;q=0.1';

// create the accept factory
$accept_factory = new AcceptFactory($_SERVER);

// create the accept object
$accept = $accept_factory->newInstance();

// assume our application has `application/json` and `text/csv` available
// as media types, in order of highest-to-lowest preference for delivery
$available = array(
    'application/json',
    'text/csv',
);

// get the best match between what the request finds acceptable and what we
// have available; the result in this case is 'text/csv'
$media = $accept->negotiateMedia($available);
echo $media->getValue(); // text/csv
?>
```

If the requested URL ends in a recognized file extension for a media type,
the _MediaNegotiator_ object used by _Accept_ will use that file extension
instead of the explicit `Accept` header value to determine the acceptable media
type:

```php
<?php
// assume the request indicates these Accept values (XML is best, then CSV,
// then anything else)
$_SERVER['HTTP_ACCEPT'] = 'application/xml;q=1.0,text/csv;q=0.5,*;q=0.1';

// assume also that the request URI explicitly notes a .json file extension
$_SERVER['REQUEST_URI'] = '/path/to/entity.json';

// create the accept factory
$accept_factory = new AcceptFactory($_SERVER);

// factory the accept object
$accept = $accept_factory->newInstance();

// assume our application has `application/json` and `text/csv` available
// as media types, in order of highest-to-lowest preference for delivery
$available = array(
    'application/json',
    'text/csv',
);

// get the best match between what the request finds acceptable and what we
// have available; the result in this case is 'application/json' because of
// the file extenstion overriding the Accept header values
$media = $accept->negotiateMedia($available);
echo $media->getValue(); // application/json
?>
```

(See the _MediaNegotiator_ class file for the list of what file extensions map
to what media types.)

To create your own mappings, set them into the _AcceptFactory_ object at
construction time:

```php
<?php
$accept_factory = new AcceptFactory($_SERVER, array(
    '.foo' => 'application/x-foo-content-type',
));

$accept = $accept_factory->newInstance();
?>
```

If the acceptable values indicate additional parameters, you can match on those
as well:

```php
<?php
// assume the request indicates these Accept values (XML is best, then CSV,
// then anything else)
$_SERVER['HTTP_ACCEPT'] = 'text/html;level=1;q=0.5,text/html;level=3';

// create the accept factory
$accept_factory = new AcceptFactory($_SERVER);

// factory the accept object
$accept = $accept_factory->newInstance();

// assume our application has these available as media types,
// in order of highest-to-lowest preference for delivery
$available = array(
    'text/html;level=1',
    'text/html;level=2',
);

// get the best match between what the request finds acceptable and what we
// have available; the result in this case is 'text/html;level=1'
$media = $accept->negotiateMedia($available);
echo $media->getValue(); // text/html
var_dump($media->getParameters()); // array('level' => '1')
?>
```

> N.b. Parameters in the acceptable values that are not present in the
> available values will not be used for matching.


### Negotiating Other Values

The other negotiation methods work much the same way, although they are less
complex than media-type negotiation.

```php
<?php
// assume the request indicates these Accept-* values
$_SERVER['HTTP_ACCEPT_CHARSET'] = 'iso-8859-5, unicode-1-1;q=0.8';
$_SERVER['HTTP_ACCEPT_ENCODING'] = 'compress;q=0.5, gzip;q=1.0';
$_SERVER['HTTP_ACCEPT_LANGUAGE'] = 'en-US, en-GB, en, *';

// create the accept factory
$accept_factory = new AcceptFactory($_SERVER);

// factory the accept object
$accept = $accept_factory->newInstance();

// charset negotiation
$available_charsets = array('iso-1234', 'unicode-1-1');
$charset = $accept->negotiateCharset($available_charsets);
echo $charset->getValue('unicode-1-1');

// encoding negotiation
$available_encodings = array();
$encoding = $accept->negotiateEncoding($available_encodings);
var_dump($encoding); // false

// language negotiation
$available_languages = array('pt-BR', 'fr-FR');
$language = $accept->negotiateLanguage($available_languages);
echo $language->getValue(); // pt-BR
?>
```


CHANGES.md
# 2.2.5
* Fixed bug when quality value was float, the array key was becoming zero.
Now converted to string https://github.com/auraphp/Aura.Accept/pull/10
* phpunit added to `require-dev` of composer. https://github.com/auraphp/Aura.Accept/pull/12


src/AbstractValues.php
<?php
/**
 *
 * This file is part of Aura for PHP.
 *
 * @package Aura.Accept
 *
 * @license http://opensource.org/licenses/bsd-license.php BSD
 *
 */
namespace Aura\Accept;

use ArrayIterator;
use Aura\Accept\Value\ValueFactory;
use IteratorAggregate;

/**
 *
 * Represents a collection of `Accept*` header values, sorted in quality order.
 *
 * @package Aura.Accept
 *
 */
abstract class AbstractValues implements IteratorAggregate
{
    /**
     *
     * An array of objects representing acceptable values from the header.
     *
     * @var array
     *
     */
    protected $acceptable = array();

    /**
     *
     * The $_SERVER key to use when populating acceptable values.
     *
     * @var string
     *
     */
    protected $server_key;

    /**
     *
     * A factory to create value objects.
     *
     * @var ValueFactory
     *
     */
    protected $value_factory;

    /**
     *
     * The type of value object to create using the ValueFactory.
     *
     * @var string
     *
     */
    protected $value_type;

    /**
     *
     * Constructor.
     *
     * @param ValueFactory $value_factory A factory for value objects.
     *
     * @param array $server A copy of $_SERVER for finding acceptable values.
     *
     */
    public function __construct(
        ValueFactory $value_factory,
        array $server = array()
    ) {
        $this->value_factory = $value_factory;
        if (isset($server[$this->server_key])) {
            $this->set($server[$this->server_key]);
        }
    }

    /**
     *
     * Returns a value object by its sorted quality position.
     *
     * @param int $key The sorted position.
     *
     * @return Value\AbstractValue
     *
     */
    public function get($key = null)
    {
        if ($key === null) {
            return $this->acceptable;
        }
        return $this->acceptable[$key];
    }

    /**
     *
     * Sets the collection to one or more acceptable values, overwriting all
     * previous values.
     *
     * @param string|array $acceptable An `Accept*` string value; e.g.,
     * `text/plain;q=0.5,text/html,text/*;q=0.1`.
     *
     * @return null
     *
     */
    protected function set($acceptable = null)
    {
        $this->acceptable = array();
        $this->add($acceptable);
    }

    /**
     *
     * Adds one or more acceptable values to this collection.
     *
     * @param string|array $acceptable One or more `Accept*` string values;
     * e.g., the string `'text/plain;q=0.5,text/html,text/*;q=0.1'` and
     * `array('text/plain;q=0.5','text/html','text/*;q=0.1')` are
     * equivalent.
     *
     * @return null
     *
     * @todo Allow this to take an array so we can parse-and-sort in one pass.
     *
     */
    protected function add($acceptable = null)
    {
        foreach ((array) $acceptable as $string) {
            $this->parse($string);
        }

        $this->sort();
    }

    /**
     *
     * Parses an acceptable string value into the `$acceptable` property.
     *
     * @param string $string An `Accept*` string value; e.g.,
     * `text/plain;q=0.5,text/html,text/*;q=0.1`.
     *
     * @return array
     *
     */
    protected function parse($string)
    {
        $acceptable = explode(',', $string);

        foreach ($acceptable as $value) {
            $pairs = explode(';', $value);
            $value = $pairs[0];
            unset($pairs[0]);

            $parameters = array();
            foreach ($pairs as $pair) {
                $param = array();
                preg_match('/^(?P<name>.+?)=(?P<quoted>"|\')?(?P<value>.*?)(?:\k<quoted>)?$/', $pair, $param);

                $parameters[$param['name']] = $param['value'];
            }

            $quality = 1.0;
            if (isset($parameters['q'])) {
                $quality = $parameters['q'];
                unset($parameters['q']);
            }

            $this->acceptable[] = $this->value_factory->newInstance(
                $this->value_type,
                trim($value),
                (float) $quality,
                $parameters
            );
        }
    }

    /**
     *
     * Sorts the `$acceptable` values according to quality levels.
     *
     * This is an unusual sort. Normally we'd think a reverse-sort would
     * order the array by q values from 1 to 0, but the problem is that
     * an implicit 1.0 on more than one value means that those values will
     * be reverse from what the header specifies, which seems unexpected
     * when negotiating later.
     *
     */
    protected function sort()
    {
        // q-value buckets
        $bucket = array();

        // sort into q-value buckets
        foreach ($this->acceptable as $value) {
            $bucket[$value->getQuality()][] = $value;
        }

        // reverse-sort the buckets so that q=1 is first and q=0 is last,
        // but the values in the buckets stay in the original order.
        krsort($bucket);

        // flatten the buckets back into the acceptable array
        $this->acceptable = array();
        foreach ($bucket as $q => $acceptable) {
            foreach ($acceptable as $value) {
                $this->acceptable[] = $value;
            }
        }
    }

    /**
     *
     * IteratorInterface: returns the iterator for this object.
     *
     * @return ArrayIterator
     *
     */
    public function getIterator()
    {
        return new ArrayIterator($this->acceptable);
    }

    /**
     *
     * Negotiates between acceptable and available values.  On success, the
     * return value is a plain old PHP object with the matching negotiated
     * `$acceptable` and `$available` value objects; these are to be inspected
     * by the calling code.
     *
     * @param array $available Available values in preference order, if any.
     *
     * @return mixed A plain-old PHP object with `$acceptable` and
     * `$available` value objects on success, or false on failure.
     *
     */
    public function negotiate(array $available = null)
    {
        // if none available, no possible match
        if (! $available) {
            return false;
        }

        // convert to object
        $clone = clone $this;
        $clone->set($available);
        $available = $clone;

        // if nothing acceptable specified, use first available
        if (! $this->acceptable) {
            return (object) array(
                'acceptable' => false,
                'available' => $available->get(0),
            );
        }

        // loop through acceptable values
        foreach ($this->acceptable as $accept) {

            // if the acceptable quality is zero, skip it
            if ($accept->getQuality() == 0) {
                continue;
            }

            // if acceptable value is "anything" return the first available
            if ($accept->isWildcard()) {
                return (object) array(
                    'acceptable' => $accept,
                    'available' => $available->get(0),
                );
            }

            // if acceptable value is available, use it
            foreach ($available as $avail) {
                if ($accept->match($avail)) {
                    return (object) array(
                        'acceptable' => $accept,
                        'available' => $avail,
                    );
                }
            }
        }

        return false;
    }
}
